### PR TITLE
Add iOS and Swift server SDK generation

### PR DESCRIPTION
**Pull request by tfrank64**
This code enables SDK generation from a swagger file. It works well with out own server generation from a swagger file. Reusing validation methods and file resource input methods.

### CODE REVIEW

**Comment by tfrank64** [comment]
In the process of fixing the build.

File: refresh/index.js
Comment (tunniclm) [inline refresh/index.js]: This is where the Swagger path should be defaulted, and the check/error makes the API explicit and more robust to change.

File: app/index.js
Comment (tunniclm) [inline app/index.js]: The actual generation should happen in the back-end generator (refresh) rather than the app generator.

File: app/index.js
Comment (tunniclm) [inline app/index.js]: Generally we use `this.log()` print warnings/messages and `this.env.error()` for fatal errors (it terminates the generator). We use `chalk` to colour error text.

File: app/index.js
Comment (tunniclm) [inline app/index.js]: The code for handling the exampleEndpoints belongs in the refresh generator, not in the prompting so that non-prompted routes can use it. Also, exampleEndpoints and a custom swagger path should be mutually exclusive (this is dealt with already in the existing code in the refresh generator).

File: app/index.js
Comment (tunniclm) [inline app/index.js]: `validateFilePathOrURL` would be a more descriptive name for this function.

File: lib/actions.js
Comment (tunniclm) [inline lib/actions.js]: I don't understand the motivation for this change, can you explain?

File: lib/helpers.js
Comment (tunniclm) [inline lib/helpers.js]: All file writes (where possible) have been using the memory file system provided by yeoman (for a couple of reasons: conflict resolver, visibility of pending writes to composed generators, rollback on abort).

File: test/integration/app/prompted_build.js
Comment (tunniclm) [inline test/integration/app/prompted_build.js]: Tests that don't require compilation should be in a separate nobuild test file and pass "--skip-build".

File: test/integration/app/prompted_nobuild.js
Comment (tunniclm) [inline test/integration/app/prompted_nobuild.js]: Seems like a rather large timeout for nobuild tests--have you been seeing issues with timeouts?

File: test/integration/app/prompted_nobuild.js
Comment (tunniclm) [inline test/integration/app/prompted_nobuild.js:336]: Are these details used in the tests? If not, I'd prefer they were not included.

**Review by tunniclm — CHANGES_REQUESTED** [review]

File: app/index.js
Comment (tfrank64) [reply app/index.js]: @tunniclm My issue is I need a reference to the example swagger file before we even get to the refresh generator. In order to generate an iOS SDK at the same time a user might generate a server SDK. Should I move things around and then have network requests happening in 2 different parts of the flow?

File: app/index.js
Comment (tfrank64) [reply app/index.js]: Ok, is there a good location to do it before any files are actually created?

File: lib/actions.js
Comment (tfrank64) [reply lib/actions.js]: So in the generated SDK use case, a user may want to create a generated server SDK from a swagger file. Since we want to incorporate this SDK into our generated server, it needs to be created before we build our `Package.swift` for example. Thus, at this point in the generator, we may have some iOS or server SDK zip files or folders in an otherwise empty directory. This `ensureEmptyDirectory` was blocking me because of those files, so I added an exception.

File: lib/helpers.js
Comment (tfrank64) [reply lib/helpers.js]: @tunniclm What do you mean exactly? Looks like `fs` is used everywhere in the generator.

File: test/integration/app/prompted_build.js
Comment (tfrank64) [reply test/integration/app/prompted_build.js]: This test does require compilation since we are adding a new module to the swift server and modifying the Package.swift, I want to ensure it compiles.

File: test/integration/app/prompted_nobuild.js
Comment (tfrank64) [reply test/integration/app/prompted_nobuild.js]: Right, so when you select the `Endpoints from swagger file` option with a Swagger file, a companion iOS SDK is created. Ricardo and I thought that was good behavior since the iOS SDK can communicate directly with the generated endpoints. A side effect is the tests take a little bit longer because we are performing a network call. Not sure if we should change generator, tests, or just let it be?

File: test/integration/app/prompted_build.js
Comment (tunniclm) [reply test/integration/app/prompted_build.js]: Yes, there should be a compile test for sure. However, the asserts below this point do not need a compilation to test, so they should be in a nobuild test so that can be run under `npm run testintfast`. Only the two asserts above this line need compiles. Sorry for not being clearer.

File: app/index.js
Comment (tunniclm) [reply app/index.js]: @tfrank64 could/should that code be moved into the refresh generator? The separation of concerns we have been using is that the app generator is responsible for the front-end work, like prompting and processing input and the refresh generator is responsible for the back-end work like content generation and file writing.

File: test/integration/app/prompted_nobuild.js
Comment (tfrank64) [reply test/integration/app/prompted_nobuild.js:336]: So when I ran the test previously, this Swagger Json was invalid. I added more details to make it valid, but we can simplify the Swagger Json if you would like? Currently, we could get rid of the `/persons` `put` endpoint and the tests will still pass, but I think thats all that can be taken away without further modification.

File: app/index.js
Comment (tunniclm) [reply app/index.js]: What we've been doing for this sort of self-contained generation work is to add it in the default phase, between `initializing` and `writing`. We've been using the prefix `build` to denote this sort of activity, for an example, see `buildSwagger` in  `refresh/index.js`.

(Aside: for logic that is closely coupled to the writing of files, we've been including it within an action inside the `writing` phase (usually prefixed with `create` or `write`). For an example, see `createCRUD` in `refresh/index.js`.)

File: lib/actions.js
Comment (tunniclm) [reply lib/actions.js]: OK. This would not be a problem if the files were written with the memory filesystem, since writes are deferred. This goes with my comment regarding the use of `fs` rather than `this.fs`.

File: lib/helpers.js
Comment (tunniclm) [reply lib/helpers.js]: Some places we use `fs` to read the filesystem (we can probably avoid this more than we have, but its not so important, especially not in the app generator since it runs first), but all the file writing (which is in the refresh generator) should be through `this.fs`.

File: app/index.js
Comment (tfrank64) [reply app/index.js]: My impression is that this is possible as long as we generate the SDKs before anything else. I can start moving this, but might take a bit more time.

File: test/integration/app/prompted_nobuild.js
Comment (tunniclm) [reply test/integration/app/prompted_nobuild.js]: OK, that could be fine provided the network call does not cause the `npm run testintfast` to take more than a minute or so to run. If it does, then we should probably either tag the slow tests or move them to another file.

File: app/index.js
Comment (tfrank64) [reply app/index.js]: Ok cool, I think most of the SDK generator work can go before the `writing` phase. I believe the Package.swift additions are already in the right spot within the `writing` phase.

File: test/integration/app/prompted_nobuild.js
Comment (tunniclm) [reply test/integration/app/prompted_nobuild.js:336]: After ending up with a mess in some of the prior tests, we've been trying to add only the minimum necessary to the test fixtures for what the test needs to run. I'm surprised that the generator doesn't fail on invalid Swagger input since I thought we ran it through a Swagger validator. The intent is to keep the test very focused. I'm fine with a bit more in the Swagger, provided it is actually needed for the test, so another option is to check we are validating properly and/or to add more assets that test we are producing the right output to correspond to this extra input data.

File: lib/helpers.js
Comment (tfrank64) [reply lib/helpers.js]: hmm ok, what exactly is the difference between `fs` and `this.fs`? Should I keep this `ingnoreFile` function in the helpers.js file?

File: test/integration/app/prompted_build.js
Comment (tfrank64) [reply test/integration/app/prompted_build.js]: Ok, that makes sense. I can move the relevant asserts to a nobuild test.

File: test/integration/app/prompted_nobuild.js
Comment (tfrank64) [reply test/integration/app/prompted_nobuild.js]: Currently, its been taking 10 or 12 seconds total, so still pretty fast.

File: test/integration/app/prompted_nobuild.js
Comment (tfrank64) [reply test/integration/app/prompted_nobuild.js:336]: That all makes sense. I'm not too familiar with this particular test, but I do know that the swagger json [here](https://github.com/IBM-Swift/generator-swiftserver/blob/master/test/integration/app/prompted_nobuild.js#L304) is not valid.

**Comment by tfrank64** [comment]
Besides travis, I think things are looking good.

Unit tests are failing due to some invalid swagger json, which seems to be fixed in one of @nhardman 's branches.

File: refresh/index.js
Comment (tunniclm) [inline refresh/index.js]: This scoping property is not necessary for actions in the default phase.

Comment (tunniclm) [reply refresh/index.js]: Also, we've typically been placing actions in the part of the file where it matches the order of execution of the phases. So it would be better if these were moved to be with the other default phase actions like `parseFromSwagger`.

File: refresh/index.js
Comment (tunniclm) [inline refresh/index.js]: `err` doesn't seem to be defined

File: refresh/index.js
Comment (tunniclm) [inline refresh/index.js]: `this.serverSwaggerFiles <= 0` looks like it should be `this.serverSwaggerFiles.length <= 0`.

**Comment by tfrank64** [comment]
@tunniclm I addressed latest review comments and tests work now. I adjusted the `mocha.opts` timeout slightly because a few of the tests cause an iOS SDK to be generated and the tests were ending too early.

**Comment by tfrank64** [comment]
@tunniclm Did you find anything else that needs to be changed? If not, can we merge this tomorrow at some point?

File: lib/actions.js
Comment (tunniclm) [reply lib/actions.js]: Is this part still necessary now that writing of the SDK files occurs later? I would prefer this check actually checks for an empty directory.

File: lib/sdkGenHelper.js
Comment (tunniclm) [inline lib/sdkGenHelper.js]: `rimraf` is not in the `dependencies` in the `package.json`. Since it *is* in the `devDependencies` already, it just needs to be moved across. The tests don't fail on this because the module can often be found by `require` due to the way modules are installed in newer versions of npm, but it probably should not be relied upon (and won't work on `npm@2` IIRC).

**Comment by tunniclm** [comment]
@tfrank64 Yes! Sorry for the delay, been working through some other defects and releases. There are a couple of things I spotted, I've added line comments for those and I'd like to see them addressed before merging.

Also, there are some other prompt UI changes I would like to see, but I don't think they are important enough to block merging this.

**Comment by tunniclm** [comment]
Things to look at before merging:

- [x] https://github.com/IBM-Swift/generator-swiftserver/pull/225/files#r123517938
- [x] https://github.com/IBM-Swift/generator-swiftserver/pull/225/files#r123519916
- [x] https://github.com/IBM-Swift/generator-swiftserver/pull/225/files#r123522429

File: refresh/templates/common/Package.swift
Comment (tunniclm) [inline refresh/templates/common/Package.swift]: Are we going to end up with 2 `Target(name: "Application", ...)` entries if we are generating a CRUD application *and* supplying `sdkTargets`?

File: lib/actions.js
Comment (tfrank64) [reply lib/actions.js]: I don't think it is necessary anymore. I will remove it.

File: refresh/templates/common/Package.swift
Comment (tfrank64) [reply refresh/templates/common/Package.swift]: I think you are right, I will see if I can reconfigure things

**Comment by tfrank64** [comment]
@tunniclm I believe I have addressed the latest code review comments. Looks like I need to rebase again though since develop was updated. Can we get this merged soon or at least hold off on merging anything else into develop before this so I can avoid a 5th rebase?

**Review by tunniclm — APPROVED** [review]

**Comment by tunniclm** [comment]
Thanks for addressing those comment, I think we should be good to merge when the testing is complete. 
Don't worry about rebasing, I will make everything good during the merge.

**EDIT**: Since the tests didn't actually start running yet, I'll just rebase now.

**Comment by tunniclm** [comment]
Had errors in the rebase, but I couldn't see why the patch in question wouldn't apply, everything looked okay to me. Just to be safe, merged `origin/develop` into this branch instead. We'll be squashing this branch into `develop` so it shouldn't make a difference.

**Comment by tunniclm** [comment]
Looks like the change to the `package.json` is causing all the integration tests to be skipped (which is why the coverage looks so bad). This is really a latent bug in the original `script` commands in the `package.json`. The paths passed to mocha to tell it which tests to run contain globs and need to be quoted so that the shell does not expand them. This is because the globs are using `**` which is not recognised by the shell glob expansion, but it is understood by the glob package used by mocha.

Long story short, all the paths supplied to mocha should be quoted. I will make that change and retest.

**Comment by tunniclm** [comment]
Given how slow and flaky our Travis tests are right now, I think we will need to addressed mocking out the `request` module so the tests can be self-contained and aren't reliant on remote services for our tests to succeed. There have been multiple intermittent build failures just in testing this PR that seem to be related to either timeouts of remote services or bad responses (complaints about invalid JSON documents).

I'm not saying we need to block merging this, but I'm adding it to the top of the list of things to address before next release (since it will impact how easily we can test and merge other pull requests).

Here's what's on that list so far:

- [x] Mock out remote network connections (`request` module)
- [x] Improve test coverage, particularly error paths
- [ ] Address standalone UI (prompt) flow

**Comment by tfrank64** [comment]
@tunniclm Yes that makes a lot of sense. The tests already take awhile testing 3 versions of Node so it will be good to address those issues.

**Comment by tunniclm** [comment]
@tfrank64 Yeah, we've been brainstorming how to speed up the builds and make them less flaky. A lot of the Travis features like caching and docker don't seem like they will address the fundamental problem of slow setup, especially on mac (mostly the Swift SDK download). For example, caching still does a download, just from S3 instead of whereever, and docker won't help us test on macOS, which is the slow platform.

If we could address the setup speed, we might be able to make some time savings by running some of the tests in parallel, but with limited concurrent jobs and long setup times I'm not sure we'll be gaining much in wall-time and we'd be adding more work for travis (duplicate setup work).

Let me know if you have any ideas.

**Pull request merged**
